### PR TITLE
Add Lightning DB core

### PR DESCRIPTION
**Pull request by jnwatson**
Only tested on Ubuntu 16.04 64-bit python 3.6 so far, but it is definitely ready to start reviewing.

### CODE REVIEW

File: synapse/cores/lmdb.py [hunk 1]
```diff
@@ -0,0 +1,564 @@
+import sys
+import msgpack
+import lmdb
+import xxhash
```
Comment (williballenthin) [inline synapse/cores/lmdb.py]: need to update `setup.py` to account for these new deps

Comment (jnwatson) [inline synapse/cores/lmdb.py]: Thanks. Done.

**Comment by jnwatson** [comment]
Damn. I done been ninja'd. You guys be busy. I'll fix it up. It looks like I need a new index.

**Comment by williballenthin** [comment]
looks like there might be a bug with non-native msgpack:

```

<ipython-input-11-3263b8a84a13> in make_cortexes()
     10         logger.info('removing existing lmdb database')
     11         os.remove(os.path.join('.', 'lmdb-bench.db'))
---> 12     lmdb_cortex = synapse.cortex.openurl('lmdb:///lmdb-bench.db')
     13 
     14 

/home/user/Documents/code/synapse/synapse/cortex.py in openurl(url, **opts)
     68 
     69     link[1].update(opts)
---> 70     return openlink(link)
     71 
     72 def openlink(link):

/home/user/Documents/code/synapse/synapse/cortex.py in openlink(link)
     78         return s_telepath.openlink(link)
     79 
---> 80     return ctor(link)
     81 
     82 def choptag(tag):

/home/user/Documents/code/synapse/synapse/cores/common.py in __init__(self, link)
    190 
    191         with self.getCoreXact() as xact:
--> 192             self._initCoreModels()
    193 
    194         # and finally, strap in our event handlers...

/home/user/Documents/code/synapse/synapse/cores/common.py in _initCoreModels(self)
    344 
    345         # now we lift/initialize from the tufos...
--> 346         for tufo in self.getTufosByProp('syn:type'):
    347             self._initTypeTufo(tufo)
    348 

/home/user/Documents/code/synapse/synapse/cores/common.py in getTufosByProp(self, prop, valu, mintime, maxtime, limit)
   1312             return self._getTufosByCache(prop,valu,limit)
   1313 
-> 1314         return self._getTufosByProp(prop, valu=valu, mintime=mintime, maxtime=maxtime, limit=limit)
   1315 
   1316     def _getTufosByProp(self, prop, valu=None, mintime=None, maxtime=None, limit=None):

/home/user/Documents/code/synapse/synapse/cores/common.py in _getTufosByProp(self, prop, valu, mintime, maxtime, limit)
   1315 
   1316     def _getTufosByProp(self, prop, valu=None, mintime=None, maxtime=None, limit=None):
-> 1317         rows = self.getJoinByProp(prop, valu=valu, mintime=mintime, maxtime=maxtime, limit=limit)
   1318         return self._rowsToTufos(rows)
   1319 

/home/user/Documents/code/synapse/synapse/cores/common.py in getJoinByProp(self, prop, valu, mintime, maxtime, limit)
   1071 
   1072         '''
-> 1073         return tuple(self._getJoinByProp(prop, valu=valu, mintime=mintime, maxtime=maxtime, limit=limit))
   1074 
   1075     def getPivotRows(self, prop, byprop, valu=None, mintime=None, maxtime=None, limit=None):

/home/user/Documents/code/synapse/synapse/cores/common.py in _getJoinByProp(self, prop, valu, mintime, maxtime, limit)
   2197 
   2198     def _getJoinByProp(self, prop, valu=None, mintime=None, maxtime=None, limit=None):
-> 2199         for irow in self._getRowsByProp(prop,valu=valu,mintime=mintime,maxtime=maxtime,limit=limit):
   2200             for jrow in self._getRowsById(irow[0]):
   2201                 yield jrow

/home/user/Documents/code/synapse/synapse/cores/lmdb.py in _getRowsByProp(self, prop, valu, limit, mintime, maxtime, do_count_only, do_delete_only)
    489                 elif not do_count_only or v_is_hashed:
    490                     # If we hashed, we must double check that val actually matches in row
--> 491                     row = self._getRowByPkValEnc(txn, value)
    492                     if v_is_hashed:
    493                         if valu != row[2]:

/home/user/Documents/code/synapse/synapse/cores/lmdb.py in _getRowByPkValEnc(self, txn, pk_val_enc, do_delete)
    321         unpacker = msgpack.Unpacker(use_list=False, encoding='utf8')
    322         unpacker.feed(row[UUID_SIZE:])
--> 323         p = unpacker.unpack()
    324         v = _dec_val_val(unpacker)
    325         t = unpacker.unpack()

/home/user/env3/lib/python3.5/site-packages/msgpack_python-0.4.8-py3.5-linux-x86_64.egg/msgpack/fallback.py in unpack(self, write_bytes)
    568 
    569     def unpack(self, write_bytes=None):
--> 570         ret = self._fb_unpack(EX_CONSTRUCT, write_bytes)
    571         self._fb_consume()
    572         return ret

/home/user/env3/lib/python3.5/site-packages/msgpack_python-0.4.8-py3.5-linux-x86_64.egg/msgpack/fallback.py in _fb_unpack(self, execute, write_bytes)
    496 
    497     def _fb_unpack(self, execute=EX_CONSTRUCT, write_bytes=None):
--> 498         typ, n, obj = self._read_header(execute, write_bytes)
    499 
    500         if execute == EX_READ_ARRAY_HEADER:

/home/user/env3/lib/python3.5/site-packages/msgpack_python-0.4.8-py3.5-linux-x86_64.egg/msgpack/fallback.py in _read_header(self, execute, write_bytes)
    345         obj = None
    346         c = self._fb_read(1, write_bytes)
--> 347         b = ord(c)
    348         if   b & 0b10000000 == 0:
    349             obj = b

TypeError: ord() expected string of length 1, but memoryview found
```

**Comment by williballenthin** [comment]
stacktrace above is in fact a msgpack issue. support for memoryviews in fallback mode has only been added since the last bugfix release, so its only available in git. i've opened an issue asking for 0.4.9 to be tagged: https://github.com/msgpack/msgpack-python/issues/222

@nic this means the lmdb cortex will require the native msgpack library until 0.4.9 comes out.

**Comment by jnwatson** [comment]
@williballenthin ty for submitting bug report to msgpack.

**Pull request merged**